KAFKA-16560: Refactor/cleanup BrokerNode/ControllerNode/ClusterConfig

brandboat opened the pull request:
https://issues.apache.org/jira/browse/KAFKA-16560
- Make ClusterConfig immutable and refactor ClusterConfig builder
- Make BrokerNode immutable and refactor out build argument
- Refactor out ControllerNode build argument
- Refactor TestKitNodes

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

brandboat committed:
Make ClusterConfig immutable

brandboat committed:
Make BrokerNode immutable

brandboat committed:
Refactor out build argument in ControllerNode

chia7712 commented on core/src/test/java/kafka/test/ClusterConfig.java:
```diff
@@ -139,28 +159,46 @@ public Map<String, String> nameTags() {
+        return builder;
+    }
+
+    private static Properties copyOf(Properties properties) {
```
Instead of creating copy, we can change the type of inner variable from `Properties` to `Map` to return unmodifiable view.

brandboat replied:
Glad to hear that, we should always prefer using Map over Properties when storing key-value pairs.

chia7712 commented on core/src/test/java/kafka/testkit/BrokerNode.java:
```diff
@@ -66,11 +68,27 @@ public Builder setNumLogDirectories(int numLogDirectories) {
+        }
+
+        public Builder setPropertyOverrides(Map<String, String> propertyOverrides) {
+            this.propertyOverrides = propertyOverrides;
```
We should do a deep copy to avoid modification from caller in the future.

chia7712 commented on core/src/test/java/kafka/test/ClusterConfig.java:
```diff
@@ -139,28 +159,46 @@ public Map<String, String> nameTags() {
 
-    public static Builder defaultClusterBuilder() {
-        return new Builder(Type.ZK, 1, 1, true, SecurityProtocol.PLAINTEXT, MetadataVersion.latestTesting());
+    public static Builder clusterBuilder() {
```
It seems to me `builder` is good enough.

chia7712 replied:
BTW, could we apply this pattern to `BrokerNode` and `ControllerNode` that we encourage users to create builder by static method?

chia7712 commented on core/src/test/java/kafka/test/ClusterConfig.java:
```diff
@@ -139,28 +159,46 @@ public Map<String, String> nameTags() {
+                .listenerName(clusterConfig.listenerName)
+                .trustStoreFile(clusterConfig.trustStoreFile)
+                .metadataVersion(clusterConfig.metadataVersion);
+        builder.serverProperties = copyOf(clusterConfig.serverProperties);
```
why we don't use setter here?

chia7712 commented on core/src/test/java/kafka/test/ClusterConfig.java:
```diff
@@ -139,28 +159,46 @@ public Map<String, String> nameTags() {
-    public static Builder clusterBuilder(Type type, int brokers, int controllers, boolean autoStart,
-                                         SecurityProtocol securityProtocol, MetadataVersion metadataVersion) {
-        return new Builder(type, brokers, controllers, autoStart, securityProtocol, metadataVersion);
+    public static Builder clusterBuilder(ClusterConfig clusterConfig) {
```
Could we add a constructor to Builder to accept a `ClusterConfig` to initialize a builder?

chia7712 commented on core/src/test/scala/unit/kafka/server/ApiVersionsRequestTest.scala:
```diff
@@ -17,48 +17,78 @@
-    new ClusterConfigProperty(key = "unstable.api.versions.enable", value = "false"),
-    new ClusterConfigProperty(key = "unstable.metadata.versions.enable", value = "true"),
+  @ClusterTests(Array(
+    new ClusterTest(clusterType = Type.ZK, metadataVersion = MetadataVersion.IBP_3_8_IV0, serverProperties = Array(
```
Pardon me. Why we need those changes?

brandboat replied:
Due to `super.brokerPropertyOverrides(config.serverProperties())` in `BeforeEach`
```java
  // Configure control plane listener to make sure we have separate listeners for testing.
  def brokerPropertyOverrides(properties: Properties): Unit = {
    if (!cluster.isKRaftTest) {
      val controlPlaneListenerName = "CONTROL_PLANE"
      val securityProtocol = cluster.config().securityProtocol()
      properties.setProperty(KafkaConfig.ControlPlaneListenerNameProp, controlPlaneListenerName)
      properties.setProperty(KafkaConfig.ListenerSecurityProtocolMapProp, s"$controlPlaneListenerName:$securityProtocol,$securityProtocol:$securityProtocol")
      properties.setProperty("listeners", s"$securityProtocol://localhost:0,$controlPlaneListenerName://localhost:0")
      properties.setProperty(KafkaConfig.AdvertisedListenersProp, s"$securityProtocol://localhost:0,${controlPlaneListenerName}://localhost:0")
    }
  }
```
Which add properties if cluster type is under ZK mode. While now we make serverProperties immutable, so I have to list all test configs for all cluster type here.

chia7712 replied:
Could we simplify the annotation by allowing to define `ClusterTemplate` in `ClusterTests`?

brandboat replied:
Currently we can't add `ClusterTemplate` in ClusterTests. Should we do that is this PR ?

chia7712 replied:
> Currently we can't add ClusterTemplate in ClusterTests. Should we do that is this PR ?

Could you file a Jira for that? Also, we can leave a TODO here

brandboat replied:
OK, let's talk about this in that JIRA. What I've wondering here is to simplify the duplicate ClusterConfigProperty here. Add `ClusterTemplate` to `ClusterTests` is a way. But maybe we can have other solutions.

brandboat replied:
gentle ping @chia7712 , Filed https://issues.apache.org/jira/browse/KAFKA-16595

chia7712 commented on core/src/test/java/kafka/test/junit/ZkClusterInvocationContext.java:
```diff
@@ -96,83 +98,7 @@ public List<Extension> getAdditionalExtensions() {
                 // However, since we cannot create this instance until we are inside the test invocation, we have
                 // to use a container class (AtomicReference) to provide this cluster object to the test itself
-
-                // This is what tests normally extend from to start a cluster, here we create it anonymously and
```
Could we keep this comment?

chia7712 commented on core/src/test/java/kafka/test/junit/ZkClusterInvocationContext.java:
```diff
@@ -349,4 +277,86 @@ public Stream<KafkaServer> servers() {
+    private static class ClusterConfigurableIntegrationHarness extends IntegrationTestHarness {
+        private ClusterConfig clusterConfig;
+
+        public ClusterConfigurableIntegrationHarness(ClusterConfig clusterConfig) {
```
`public` -> `private`

chia7712 reviewed (commented):
@brandboat thanks for this nice refactor. a couple of comments left. please feel free to ignore them if you are going to do more updates before reviews.

brandboat committed:
Address commits

chia7712 commented on core/src/test/java/kafka/test/ClusterConfig.java:
```diff
@@ -67,13 +69,16 @@ public class ClusterConfig {
+        this.adminClientProperties = Collections.unmodifiableMap(adminClientProperties);
+        this.saslServerProperties = Collections.unmodifiableMap(saslServerProperties);
+        this.saslClientProperties = Collections.unmodifiableMap(saslClientProperties);
+        this.perBrokerOverrideProperties = Collections.unmodifiableMap(
```
In order to keep consistency, please do deep copy in eight Builder or this constructor

brandboat replied:
Pardon me, I'm not quite clear about this comment. Could you explain more ? :smiley: 
For convenience, this pr add something like `public Builder putServerProperty(String key, String value) ` in each configurations like `serverProperties`, `consumerProperties`. So we don't need to do deep copy in ClusterConfig.Builder. 
Or did you mean we need to use `setServerProperty(Map<String, String> serverProperties)`  instead of `putServerProperty(String key, String value)`  ?

chia7712 replied:
The purpose of this PR is to make `ClusterConfig` be immutable. However, `ClusterConfig` have only immutable "view" of `serverProperties`, `producerProperties`, etc. It means `ClusterConfig` is still mutable object since callers can change inner variables of `ClusterConfig` by updating `ClusterConfig#Builder`. For example, user can call `putSaslServerProperty` to chagne `saslServerProperties` of `ClusterConfig`

brandboat replied:
Got it, thanks for explanation.

chia7712 commented on core/src/test/java/kafka/test/ClusterConfig.java:
```diff
@@ -211,13 +186,36 @@ public static class Builder {
+            this.saslClientProperties = new HashMap<>(clusterConfig.saslClientProperties);
+            Map<Integer, Map<String, String>> perBrokerOverrideProps = new HashMap<>();
+            clusterConfig.perBrokerOverrideProperties.forEach((k, v) -> perBrokerOverrideProps.put(k, new HashMap<>(v)));
+            this.perBrokerOverrideProperties = perBrokerOverrideProps;
```
```java
            this.perBrokerOverrideProperties = clusterConfig.perBrokerOverrideProperties.entrySet().stream()
                    .collect(Collectors.toMap(Map.Entry::getKey, e -> new HashMap<>(e.getValue())));
```

chia7712 commented on core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java:
```diff
@@ -290,7 +287,7 @@ public void waitForReadyBrokers() throws InterruptedException {
 
         @Override
-        public void rollingBrokerRestart() {
+        public void rollingBrokerRestart(Optional<ClusterConfig> clusterConfig) {
```
As not all implementation support this method, we should remove it from interface. The callers can use `getUnderlying` to get zk instance and call that method

chia7712 commented on core/src/test/java/kafka/test/ClusterConfig.java:
```diff
@@ -22,11 +22,13 @@
-import java.util.Properties;
+import java.util.stream.Collectors;
 
 /**
```
please add comments to say this class is a immutable object

chia7712 commented on core/src/test/java/kafka/testkit/BrokerNode.java:
```diff
@@ -66,11 +74,27 @@ public Builder setNumLogDirectories(int numLogDirectories) {
-            Uuid clusterId,
-            boolean combined
-        ) {
+        public Builder setClusterId(Uuid clusterId) {
```
The builder of `ClusterConfig` do not have `set` prefix. Could we unify the naming?

chia7712 commented on core/src/test/java/kafka/testkit/TestKitNodes.java:
```diff
@@ -97,14 +96,32 @@ public Builder setBrokerNodes(int numBrokerNodes, int disksPerBroker) {
+                    throw new RuntimeException("Broker id " + brokerId + " does not exist");
+                }
+                Map<String, String> propertiesOverride = new HashMap<>();
+                properties.forEach((key, value) -> propertiesOverride.put(key.toString(), value.toString()));
```
`toString` is redundant.

chia7712 commented on core/src/test/java/kafka/testkit/TestKitNodes.java:
```diff
@@ -97,14 +96,32 @@ public Builder setBrokerNodes(int numBrokerNodes, int disksPerBroker) {
         }
 
+        /**
+         * Set per broker properties overrides, this setter must be invoked after setBrokerNodes which
```
The happen-before is anti-pattern in builder pattern. Could we merge `PerBrokerPropertiesOverrides` to `setNumControllerNodes/setNumBrokerNodes`? for example:

```java
setBrokerNodes(int numBrokerNodes, int disksPerBroker, Function<Integer, Map<String, Object>> perBrokerConfigs)

```

chia7712 replied:
It seems `TestKitNodes` need to be refactor also. In short, it should verify/apply all settings when building.

chia7712 commented on core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala:
```diff
@@ -435,15 +443,13 @@ class KRaftClusterTest {
-                                      brokerPropertyOverrides: (TestKitNodes, BrokerNode) => Map[String, String])
+                                      brokerPropertyOverrides: util.Map[Integer, util.Map[String, String]])
                                      (action: KafkaClusterTestKit => Unit): Unit = {
     val nodes = new TestKitNodes.Builder()
```
Could we change the style to `TestKitNodes.builder()` so as to unify the builder pattern in testing?

chia7712 commented on core/src/test/scala/unit/kafka/server/SaslApiVersionsRequestTest.scala:
```diff
@@ -18,41 +18,58 @@ package kafka.server
   @BeforeEach
-  def setupSasl(config: ClusterConfig): Unit = {
+  def setupSasl(): Unit = {
     sasl = new SaslSetup() {}
```
Could we create `sasl` in initialization?

brandboat replied:
Currently, no.  Since we clear `java.security.auth.login.config` in [QuorumTestHarness#tearDown](https://github.com/apache/kafka/blob/59c781415fc37c89aa087d7c2999cec7f82f6188/core/src/test/scala/integration/kafka/server/QuorumTestHarness.scala#L440). So if we initialize sasl in `@BeforeAll`, first test passed, second one raise error as below. 
```text
Could not find a 'KafkaServer' or 'control_plane.KafkaServer' entry in the JAAS configuration. System property 'java.security.auth.login.config' is not set
```
This would require some effort if we want to make this happen.

chia7712 reviewed (commented):
@brandboat I do love this PR so there is a bunch comments. please take a look.

brandboat committed:
Add setPrefix and replace put property with set map in ClusterConfig

brandboat committed:
Remove rollingBrokerRestart from ClusterInstance interface

brandboat committed:
Refactor KRaftClusterTest#doOnStartedKafkaCluster

brandboat committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-16560

chia7712 commented on core/src/test/java/kafka/test/ClusterConfig.java:
```diff
@@ -173,63 +199,104 @@ public static class Builder {
+        }
+
+        public Builder setPerBrokerProperties(Map<Integer, Map<String, String>> perBrokerOverrideProperties) {
+            this.perBrokerOverrideProperties = Collections.unmodifiableMap(
```
```java
            this.perBrokerOverrideProperties = Collections.unmodifiableMap(
                    perBrokerOverrideProperties.entrySet().stream()
                            .collect(Collectors.toMap(Map.Entry::getKey, e -> Collections.unmodifiableMap(new HashMap<>(e.getValue())))));
```

chia7712 commented on core/src/test/java/kafka/test/ClusterConfig.java:
```diff
@@ -173,63 +199,104 @@ public static class Builder {
         private String listenerName;
         private File trustStoreFile;
         private MetadataVersion metadataVersion;
+        private Map<String, String> serverProperties = Collections.unmodifiableMap(new HashMap<>());
```
Is `Collections.emptyMap();`  more suitable?

brandboat replied:
Yes. The code here is redundant. Thanks

chia7712 commented on core/src/test/java/kafka/test/ClusterConfig.java:
```diff
@@ -173,63 +199,104 @@ public static class Builder {
+
         public ClusterConfig build() {
-            return new ClusterConfig(type, brokers, controllers, name, autoStart, securityProtocol, listenerName, trustStoreFile, metadataVersion);
+            return new ClusterConfig(type, brokers, controllers, name, autoStart, securityProtocol, listenerName,
```
Could you please add null check for the non-null variables?

chia7712 commented on core/src/test/java/kafka/testkit/BrokerNode.java:
```diff
@@ -130,7 +154,7 @@ public BrokerNode build(
         this.initialMetaPropertiesEnsemble = initialMetaPropertiesEnsemble;
         this.combined = combined;
-        this.propertyOverrides = new HashMap<>(propertyOverrides);
+        this.propertyOverrides = Collections.unmodifiableMap(propertyOverrides);
```
`ClusterConfig` address the "immutable" by creating immutable objects in Builder. Could we align that behavior?

chia7712 commented on core/src/test/java/kafka/testkit/BrokerNode.java:
```diff
@@ -41,6 +46,9 @@ public static class Builder {
         private int numLogDirectories = 1;
         private String metadataDirectory = null;
         private Map<String, String> propertyOverrides = new HashMap<>();
```
`Collections.emptyMap()`

chia7712 reviewed (commented):
@brandboat please take a look at following comments. Thanks!

brandboat committed:
Address comments

chia7712 commented on core/src/test/java/kafka/testkit/BrokerNode.java:
```diff
@@ -66,11 +74,27 @@ public Builder setNumLogDirectories(int numLogDirectories) {
+        }
+
+        public Builder setPropertyOverrides(Map<String, String> propertyOverrides) {
+            this.propertyOverrides = Collections.unmodifiableMap(propertyOverrides);
```
Maybe we need to do a copy in order to avoid changes produced from outer.

brandboat replied:
Thanks, I'll address this in other builders also.

chia7712 commented on core/src/test/java/kafka/testkit/TestKitNodes.java:
```diff
@@ -97,14 +96,31 @@ public Builder setBrokerNodes(int numBrokerNodes, int disksPerBroker) {
+                if (!brokerNodeBuilders.containsKey(brokerId)) {
+                    throw new RuntimeException("Broker id " + brokerId + " does not exist");
+                }
+                Map<String, String> propertiesOverride = new HashMap<>(properties);
```
the deep copy should be addressed by `setPropertyOverrides`

chia7712 commented on core/src/test/java/kafka/test/ClusterConfig.java:
```diff
@@ -126,8 +138,12 @@ public MetadataVersion metadataVersion() {
 
-    public Properties brokerServerProperties(int brokerId) {
-        return perBrokerOverrideProperties.computeIfAbsent(brokerId, __ -> new Properties());
+    public Map<String, String> brokerServerProperties(int brokerId) {
```
This helper is used by `ZkClusterInvocationContext` only. Maybe we can remove this one to simplify `ClusterConfig`

chia7712 reviewed (commented):
@brandboat Sorry that I leave more comments below. please take a look. thanks

brandboat committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-16560

brandboat committed:
Fix config changed place

brandboat committed:
Address comments

brandboat commented:
> @brandboat Sorry that I leave more comments below. please take a look. thanks

Thank you for your patience! I appreciate your input. Already addressed all comments as above. Please take another look when you are available :smiley:

brandboat committed:
Remove redundan map copy

brandboat committed:
Add rnn in TestKitNodes

chia7712 commented on core/src/test/java/kafka/testkit/TestKitNodes.java:
```diff
@@ -97,9 +107,12 @@ public Builder setBrokerNodes(int numBrokerNodes, int disksPerBroker) {
                 if (!brokerNodeBuilders.isEmpty()) {
                     nextId = brokerNodeBuilders.lastKey() + 1;
                 }
-                BrokerNode.Builder brokerNodeBuilder = new BrokerNode.Builder()
```
Can we evaluate those settings in `build` method? That can simplify code since we don't need to revert the changes (for example, the number of brokers).

chia7712 commented on core/src/test/java/kafka/testkit/TestKitNodes.java:
```diff
@@ -167,11 +186,11 @@ private TestKitNodes(
+        this.baseDirectory = Objects.requireNonNull(baseDirectory);
+        this.clusterId = Objects.requireNonNull(clusterId);
+        this.bootstrapMetadata = Objects.requireNonNull(bootstrapMetadata);
+        this.controllerNodes = Objects.requireNonNull(controllerNodes);
```
Please follow the rule of this PR - be a immutable object

chia7712 commented on core/src/test/java/kafka/testkit/BrokerNode.java:
```diff
@@ -121,16 +146,16 @@ public BrokerNode build(
     ) {
-        this.incarnationId = incarnationId;
-        this.initialMetaPropertiesEnsemble = initialMetaPropertiesEnsemble;
+        this.incarnationId = Objects.requireNonNull(incarnationId);
```
this is not used so it is fine to remove it

chia7712 commented on core/src/test/java/kafka/testkit/BrokerNode.java:
```diff
@@ -121,16 +146,16 @@ public BrokerNode build(
         Map<String, String> propertyOverrides
     ) {
-        this.incarnationId = incarnationId;
-        this.initialMetaPropertiesEnsemble = initialMetaPropertiesEnsemble;
```
`logDataDirectories` should return immutable collection

brandboat replied:
I make `logDataDirectories()` return `Set<String>` instead `List<String>` to get rid of the redundant arraylist wrapping.

chia7712 reviewed (commented):
@brandboat thanks for updated PR. a couple of comments left.

brandboat committed:
Address comments

chia7712 commented on core/src/test/java/kafka/testkit/TestKitNodes.java:
```diff
@@ -59,73 +65,64 @@ public Builder setCombined(boolean combined) {
-        }
 
-        public TestKitNodes build() {
             String baseDirectory = TestUtils.tempDirectory().getAbsolutePath();
```
We don't need to delete `baseDirectory` since `TestUtils.tempDirectory()` will delete the return folder when terminating.

chia7712 commented on core/src/test/java/kafka/testkit/BrokerNode.java:
```diff
@@ -66,17 +69,30 @@ public Builder setNumLogDirectories(int numLogDirectories) {
-            if (incarnationId == null) {
-                incarnationId = Uuid.randomUuid();
-            }
             List<String> logDataDirectories = IntStream
```
Could you add null check for `baseDirectory`?

chia7712 commented on core/src/test/java/kafka/testkit/TestKitNodes.java:
```diff
@@ -59,73 +65,64 @@ public Builder setCombined(boolean combined) {
+                        .build();
+                    controllerNodes.put(node.id(), node);
                 }
                 TreeMap<Integer, BrokerNode> brokerNodes = new TreeMap<>();
```
We can leverage lambda here.
```java
                Map<Integer, BrokerNode> brokerNodes = IntStream.range(startBrokerId(), startBrokerId() + numBrokerNodes)
                        .boxed().collect(Collectors.toMap(Function.identity(), id -> BrokerNode.builder()
                        .setId(id)
                        .setNumLogDirectories(numDisksPerBroker)
                        .setBaseDirectory(baseDirectory)
                        .setClusterId(clusterId)
                        .setCombined(combined)
                        .setPropertyOverrides(perBrokerProperties.getOrDefault(id, Collections.emptyMap()))
                        .build()));
```

chia7712 commented on core/src/test/java/kafka/testkit/TestKitNodes.java:
```diff
@@ -59,73 +65,64 @@ public Builder setCombined(boolean combined) {
                 if (clusterId == null) {
                     clusterId = Uuid.randomUuid();
                 }
                 TreeMap<Integer, ControllerNode> controllerNodes = new TreeMap<>();
```
ditto. please try to use lambda to simplify code.

chia7712 commented on core/src/test/java/kafka/testkit/TestKitNodes.java:
```diff
@@ -167,11 +164,11 @@ private TestKitNodes(
+        this.baseDirectory = Objects.requireNonNull(baseDirectory);
+        this.clusterId = Objects.requireNonNull(clusterId);
+        this.bootstrapMetadata = Objects.requireNonNull(bootstrapMetadata);
+        this.controllerNodes = new TreeMap<>(Objects.requireNonNull(controllerNodes));
```
please set immutable wrapper

chia7712 commented on core/src/test/java/kafka/testkit/TestKitNodes.java:
```diff
@@ -167,11 +164,11 @@ private TestKitNodes(
+        this.clusterId = Objects.requireNonNull(clusterId);
+        this.bootstrapMetadata = Objects.requireNonNull(bootstrapMetadata);
+        this.controllerNodes = new TreeMap<>(Objects.requireNonNull(controllerNodes));
+        this.brokerNodes = new TreeMap<>(Objects.requireNonNull(brokerNodes));
```
ditto

chia7712 commented on core/src/test/java/kafka/testkit/TestKitNodes.java:
```diff
@@ -167,11 +164,11 @@ private TestKitNodes(
         NavigableMap<Integer, ControllerNode> controllerNodes,
         NavigableMap<Integer, BrokerNode> brokerNodes
```
We should use `Map` by default. please check the usage.

chia7712 commented on core/src/test/scala/unit/kafka/server/ApiVersionsRequestTest.scala:
```diff
@@ -84,17 +125,38 @@ class ApiVersionsRequestTest(cluster: ClusterInstance) extends AbstractApiVersio
-  @ClusterTest(metadataVersion = MetadataVersion.IBP_3_7_IV4, serverProperties = Array(
-    new ClusterConfigProperty(key = "unstable.api.versions.enable", value = "false"),
-    new ClusterConfigProperty(key = "unstable.metadata.versions.enable", value = "false"),
+  @ClusterTests(Array(
```
please add TODO and KAFKA-xxxx

brandboat replied:
Already added the TODO in the top of this file. L30

chia7712 reviewed (commented):
@brandboat thanks for updated PR. a couple of comments left.

brandboat committed:
Address comments

chia7712 commented on core/src/test/java/kafka/testkit/TestKitNodes.java:
```diff
@@ -198,7 +177,7 @@ public BootstrapMetadata bootstrapMetadata() {
     }
 
-    public NavigableMap<Integer, BrokerNode> brokerNodes() {
+    public Map<Integer, BrokerNode> brokerNodes() {
```
sorry that we need to use `NavigableMap` here since we instantiate broker/controller according to id.

chia7712 replied:
please ignore this comment since we create all controllers before brokers, and hence the order of brokers is not matter.

brandboat replied:
I found all tests  failed... Sorry for wasting your time, I'll be more careful :disappointed:

chia7712 commented on core/src/test/java/kafka/testkit/TestKitNodes.java:
```diff
@@ -59,87 +63,66 @@ public Builder setCombined(boolean combined) {
+                    .setNumLogDirectories(numDisksPerBroker)
+                    .setBaseDirectory(baseDirectory)
+                    .setClusterId(clusterId)
+                    .setCombined(combined)
```
As the number of brokers can be different to number of controllers, the `combined` should be replaced by `controllerNodes.containsKey(id)`.

chia7712 reviewed (commented):
@brandboat thanks for your updates. there are two major comments. PTAL

brandboat committed:
Fix error

brandboat committed:
Remove unused import

brandboat commented:
Failed tests seems related to this pr, I'll check them later.

chia7712 commented:
@brandboat Could you rebase code to trigger QA again? I run those tests on my local. They pass

brandboat committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-16560

brandboat committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-16560

brandboat committed:
Fix missing TreeMap copy

brandboat committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-16560

chia7712 commented:
@brandboat could you please fix the conflicts?

brandboat committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-16560

chia7712 commented on core/src/test/java/kafka/test/ClusterConfig.java:
```diff
@@ -139,28 +151,38 @@ public Map<String, String> nameTags() {
-    public static Builder clusterBuilder(Type type, int brokers, int controllers, boolean autoStart,
-                                         SecurityProtocol securityProtocol, MetadataVersion metadataVersion) {
-        return new Builder(type, brokers, controllers, autoStart, securityProtocol, metadataVersion);
+    public static Builder builder(ClusterConfig clusterConfig) {
```
Could you please add unit test for this method? That deep copy is error-prone.

brandboat replied:
> That deep copy is error-prone.

Did you mean we shouldn't send ClusterConfig as an argument to `builder` ? If so, I can remove this method (i.e. L168). But that means we have to explicitly write all setters if we want to copy a ClusterConfig to a brand new one. 

And I'm not sure why this is error-prone. Would be appreciate if you can share more 😃 .

chia7712 replied:
> Did you mean we shouldn't send ClusterConfig as an argument to builder ? If so, I can remove this method (i.e. L168). But that means we have to explicitly write all setters if we want to copy a ClusterConfig to a brand new one.
And I'm not sure why this is error-prone. Would be appreciate if you can share more 😃 .

that is a good helper method but we could neglect this copy-all when adding new field to `ClusterConfig` :(

brandboat replied:
Got it. Maybe we should directly remove the helper method here. WDYT ?

chia7712 replied:
> Maybe we should directly remove the helper method here

It seems to me this helper could be useful to the tests which have to update a part of configs. Hence, all we need to do is add test to make sure the clone works well :)

brandboat replied:
Roger that. Thanks for your patience

chia7712 commented on core/src/test/scala/integration/kafka/zk/ZkMigrationIntegrationTest.scala:
```diff
@@ -57,16 +57,18 @@ import org.slf4j.{Logger, LoggerFactory}
-    props.setProperty("listeners", "PLAINTEXT://localhost:0,EXTERNAL://localhost:0")
-    props.setProperty("advertised.listeners", "PLAINTEXT://localhost:0,EXTERNAL://localhost:0")
-    props.setProperty("listener.security.protocol.map", "EXTERNAL:PLAINTEXT,PLAINTEXT:PLAINTEXT")
+  def addZkBrokerProps(builder: ClusterConfig.Builder): Unit = {
```
It seems this method is used by `zkClustersForAllMigrationVersions` only. We can merge them into one method. For example:
```scala
      val serverProperties = new util.HashMap[String, String]()
      serverProperties.put("inter.broker.listener.name", "EXTERNAL")
      serverProperties.put("listeners", "PLAINTEXT://localhost:0,EXTERNAL://localhost:0")
      serverProperties.put("advertised.listeners", "PLAINTEXT://localhost:0,EXTERNAL://localhost:0")
      serverProperties.put("listener.security.protocol.map", "EXTERNAL:PLAINTEXT,PLAINTEXT:PLAINTEXT")
      clusterGenerator.accept(ClusterConfig.defaultBuilder()
        .setMetadataVersion(mv)
        .setBrokers(3)
        .setType(Type.ZK)
        .setServerProperties(serverProperties).build())
```

chia7712 commented on core/src/test/scala/integration/kafka/zk/ZkMigrationIntegrationTest.scala:
```diff
@@ -187,11 +188,16 @@ class ZkMigrationIntegrationTest {
-      zkCluster.config().serverProperties().put(KafkaConfig.ControllerListenerNamesProp, "CONTROLLER")
-      zkCluster.config().serverProperties().put(SocketServerConfigs.LISTENER_SECURITY_PROTOCOL_MAP_CONFIG, "CONTROLLER:PLAINTEXT,EXTERNAL:PLAINTEXT,PLAINTEXT:PLAINTEXT")
-      zkCluster.rollingBrokerRestart() // This would throw if authorizers weren't allowed
+      val serverProperties = new util.HashMap[String, String]()
```
```java
      val serverProperties = new util.HashMap[String, String](zkCluster.config().serverProperties())
```

chia7712 commented on core/src/test/scala/integration/kafka/zk/ZkMigrationIntegrationTest.scala:
```diff
@@ -319,11 +325,16 @@ class ZkMigrationIntegrationTest {
-      zkCluster.config().serverProperties().put(SocketServerConfigs.LISTENER_SECURITY_PROTOCOL_MAP_CONFIG, "CONTROLLER:PLAINTEXT,EXTERNAL:PLAINTEXT,PLAINTEXT:PLAINTEXT")
-      zkCluster.rollingBrokerRestart()
+      val serverProperties = new util.HashMap[String, String]()
+      serverProperties.putAll(zkCluster.config().serverProperties())
```
ditto

chia7712 commented on core/src/test/scala/integration/kafka/zk/ZkMigrationIntegrationTest.scala:
```diff
@@ -453,11 +464,16 @@ class ZkMigrationIntegrationTest {
-      zkCluster.config().serverProperties().put(SocketServerConfigs.LISTENER_SECURITY_PROTOCOL_MAP_CONFIG, "CONTROLLER:PLAINTEXT,EXTERNAL:PLAINTEXT,PLAINTEXT:PLAINTEXT")
-      zkCluster.rollingBrokerRestart()
+      val serverProperties = new util.HashMap[String, String]()
+      serverProperties.putAll(zkCluster.config().serverProperties())
```
ditto

chia7712 replied:
please fix this

chia7712 commented on core/src/test/scala/integration/kafka/zk/ZkMigrationIntegrationTest.scala:
```diff
@@ -517,11 +533,16 @@ class ZkMigrationIntegrationTest {
-      zkCluster.config().serverProperties().put(SocketServerConfigs.LISTENER_SECURITY_PROTOCOL_MAP_CONFIG, "CONTROLLER:PLAINTEXT,EXTERNAL:PLAINTEXT,PLAINTEXT:PLAINTEXT")
-      zkCluster.rollingBrokerRestart() // This would throw if authorizers weren't allowed
+      val serverProperties = new util.HashMap[String, String]()
+      serverProperties.putAll(zkCluster.config().serverProperties())
```
ditto

chia7712 replied:
please fix this

chia7712 commented on core/src/test/scala/integration/kafka/zk/ZkMigrationIntegrationTest.scala:
```diff
@@ -602,11 +623,16 @@ class ZkMigrationIntegrationTest {
-      zkCluster.config().serverProperties().put(SocketServerConfigs.LISTENER_SECURITY_PROTOCOL_MAP_CONFIG, "CONTROLLER:PLAINTEXT,EXTERNAL:PLAINTEXT,PLAINTEXT:PLAINTEXT")
-      zkCluster.rollingBrokerRestart()
+      val serverProperties = new util.HashMap[String, String]()
+      serverProperties.putAll(zkCluster.config().serverProperties())
```
ditto

chia7712 replied:
please fix this

chia7712 commented on core/src/test/scala/integration/kafka/zk/ZkMigrationIntegrationTest.scala:
```diff
@@ -667,11 +693,16 @@ class ZkMigrationIntegrationTest {
-      zkCluster.config().serverProperties().put(SocketServerConfigs.LISTENER_SECURITY_PROTOCOL_MAP_CONFIG, "CONTROLLER:PLAINTEXT,EXTERNAL:PLAINTEXT,PLAINTEXT:PLAINTEXT")
-      zkCluster.rollingBrokerRestart()
+      val serverProperties = new util.HashMap[String, String]()
+      serverProperties.putAll(zkCluster.config().serverProperties())
```
ditto

chia7712 replied:
please fix this

chia7712 commented on core/src/test/scala/integration/kafka/zk/ZkMigrationIntegrationTest.scala:
```diff
@@ -727,11 +758,16 @@ class ZkMigrationIntegrationTest {
-      zkCluster.config().serverProperties().put(SocketServerConfigs.LISTENER_SECURITY_PROTOCOL_MAP_CONFIG, "CONTROLLER:PLAINTEXT,EXTERNAL:PLAINTEXT,PLAINTEXT:PLAINTEXT")
-      zkCluster.rollingBrokerRestart()
+      val serverProperties = new util.HashMap[String, String]()
+      serverProperties.putAll(zkCluster.config().serverProperties())
```
ditto

chia7712 replied:
please fix this

chia7712 commented on core/src/test/scala/integration/kafka/zk/ZkMigrationIntegrationTest.scala:
```diff
@@ -802,11 +838,16 @@ class ZkMigrationIntegrationTest {
-      zkCluster.config().serverProperties().put(SocketServerConfigs.LISTENER_SECURITY_PROTOCOL_MAP_CONFIG, "CONTROLLER:PLAINTEXT,EXTERNAL:PLAINTEXT,PLAINTEXT:PLAINTEXT")
-      zkCluster.rollingBrokerRestart()
+      val serverProperties = new util.HashMap[String, String]()
+      serverProperties.putAll(zkCluster.config().serverProperties())
```
ditto

chia7712 replied:
please fix this

chia7712 commented on core/src/test/java/kafka/test/ClusterTestExtensionsTest.java:
```diff
@@ -48,14 +51,18 @@ public class ClusterTestExtensionsTest {
+        serverProperties.put("foo", "bar");
+        clusterGenerator.accept(ClusterConfig.defaultBuilder()
+                .setName("Generated Test")
+                .setServerProperties(serverProperties)
```
We can use `Collections.singletonMap("foo", "bar")`

chia7712 commented on core/src/test/java/kafka/test/ClusterTestExtensionsTest.java:
```diff
@@ -48,14 +51,18 @@ public class ClusterTestExtensionsTest {
 
     // BeforeEach run after class construction, but before cluster initialization and test invocation
     @BeforeEach
     public void beforeEach(ClusterConfig config) {
```
As `ClusterConfig` becomes immutable object, developers can modify the configs by injection of `ClusterConfig`. Maybe we should remove this usage from code base.  WDYT?

brandboat replied:
Make sense to me. Thanks

chia7712 replied:
Could you please file a ticket for it?

brandboat replied:
filed https://issues.apache.org/jira/browse/KAFKA-16627

chia7712 reviewed (commented):
@brandboat thanks for updated PR. a couple of comments left. PTAL

brandboat committed:
Address comments

chia7712 commented on core/src/test/java/kafka/test/ClusterConfigTest.java:
```diff
@@ -0,0 +1,93 @@
+public class ClusterConfigTest {
+
+    @Test
+    public void testClusterConfigBuilder() throws IOException {
```
`testCopy`

chia7712 commented on core/src/test/java/kafka/test/ClusterConfigTest.java:
```diff
@@ -0,0 +1,93 @@
+                .setPerBrokerProperties(Collections.singletonMap(0, Collections.singletonMap("broker_0", "broker_0_value")))
+                .build();
+
+        Assertions.assertEquals(Type.KRAFT, clusterConfig.clusterType());
```
Dear friend, this test can't protect us from overlooking the field copy since we don't check the field added in the future, right? Maybe we can use reflection to get all fields and then check all dynamically.
```java
    private static Map<String, Object> fields(ClusterConfig config) {
        return Arrays.stream(config.getClass().getDeclaredFields()).collect(Collectors.toMap(Field::getName, f -> {
            f.setAccessible(true);
            return Assertions.assertDoesNotThrow(() -> f.get(config));
        }));
    }

    @Test
    public void testClusterConfigBuilder() throws IOException {
        File trustStoreFile = TestUtils.tempFile();

        ClusterConfig clusterConfig = ClusterConfig.builder()
                .setType(Type.KRAFT)
                .setBrokers(3)
                .setControllers(2)
                .setName("builder-test")
                .setAutoStart(true)
                .setSecurityProtocol(SecurityProtocol.PLAINTEXT)
                .setListenerName("EXTERNAL")
                .setTrustStoreFile(trustStoreFile)
                .setMetadataVersion(MetadataVersion.IBP_0_8_0)
                .setServerProperties(Collections.singletonMap("broker", "broker_value"))
                .setConsumerProperties(Collections.singletonMap("consumer", "consumer_value"))
                .setProducerProperties(Collections.singletonMap("producer", "producer_value"))
                .setAdminClientProperties(Collections.singletonMap("admin_client", "admin_client_value"))
                .setSaslClientProperties(Collections.singletonMap("sasl_client", "sasl_client_value"))
                .setSaslServerProperties(Collections.singletonMap("sasl_server", "sasl_server_value"))
                .setPerBrokerProperties(Collections.singletonMap(0, Collections.singletonMap("broker_0", "broker_0_value")))
                .build();

        Map<String, Object> fields = fields(clusterConfig);
        Map<String, Object> copy = fields(ClusterConfig.builder(clusterConfig).build());
        Assertions.assertEquals(fields, copy);
    }
```

brandboat replied:
Sorry, I missed the important part in this test. I'm going to override `equals()` and `hashCode` in ClusterConfig instead of using reflect method.

brandboat replied:
Sign... But we may still overlook add fields in `equals()` and `hashCode`. I'll address the test as you provided in this  comment. Thanks

chia7712 commented on core/src/test/scala/integration/kafka/server/KafkaServerKRaftRegistrationTest.scala:
```diff
@@ -71,11 +72,16 @@ class KafkaServerKRaftRegistrationTest {
-      zkCluster.config().serverProperties().put(KafkaConfig.ControllerListenerNamesProp, "CONTROLLER")
-      zkCluster.config().serverProperties().put(SocketServerConfigs.LISTENER_SECURITY_PROTOCOL_MAP_CONFIG, "CONTROLLER:PLAINTEXT,EXTERNAL:PLAINTEXT,PLAINTEXT:PLAINTEXT")
-      zkCluster.rollingBrokerRestart()
+      val serverProperties = new java.util.HashMap[String, String]()
```
```java
val serverProperties = new util.HashMap[String, String](zkCluster.config().serverProperties(zkCluster.config().serverProperties()))   
```

chia7712 commented on core/src/test/scala/integration/kafka/server/KafkaServerKRaftRegistrationTest.scala:
```diff
@@ -108,11 +114,16 @@ class KafkaServerKRaftRegistrationTest {
-      zkCluster.config().serverProperties().put(SocketServerConfigs.LISTENER_SECURITY_PROTOCOL_MAP_CONFIG, "CONTROLLER:PLAINTEXT,PLAINTEXT:PLAINTEXT")
-      assertThrows(classOf[IllegalArgumentException], () => zkCluster.rollingBrokerRestart())
+      val serverProperties = new java.util.HashMap[String, String]()
+      serverProperties.putAll(zkCluster.config().serverProperties())
```
ditto

chia7712 reviewed (commented):
@brandboat thanks for updated PR. a couple of comments left.

brandboat committed:
Address comments

brandboat committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-16560

chia7712 reviewed (approved):
LGTM
